apps/github-cascading-app/src/cascading/cascading.ts

```diff
@@ -0,0 +1,412 @@
+  /**
+   * Determine of the triggered checks of a pull request are passed
+   *
+   * @param id ID of the pull request
+   */
+  protected abstract areAllChecksPassed(id: string | number, conclusion: CheckConclusion): Promise<boolean>;
+  /**
+   * Trigger the pull request merge
+   *
+   * @param id ID of the pull request
+   */
+  protected abstract mergePullRequest(id: string | number): Promise<boolean>;
+  /**
+   * Determine if the base branch is ahead of the target branch
+   *
+   * @param baseBranch
+   * @param targetBranch
+   */
+  protected abstract isBranchAhead(baseBranch: string, targetBranch: string): Promise<boolean>;
+
+  /**
+   * @param logger Logger
+   * @param username User name used for git commands
+   * @param email Email used for git commands
+   */
+  constructor(public logger: BaseLogger, public username: string = 'Auto Cascading', public email: string = 'cascading@otter.com') {
+  }
+
+  /**
+   * Retrieve the ordered cascading branches from the repository full branch list
+   *
+   * @param fullBranchesList full list of branches of the repository
+   * @param config Cascading configuration
+   */
+  protected getOrderedCascadingBranches(fullBranchesList: string[], config: CascadingConfiguration) {
+    const cascadingBranchesPattern = new RegExp(config.cascadingBranchesPattern);
+    const versionCapturePattern = new RegExp(config.versionCapturePattern);
+    const ignoredPatterns = config.ignoredPatterns.map((pattern) => new RegExp(pattern));
+    this.logger.debug('Sort the full list of branches:');
+    this.logger.debug(JSON.stringify(fullBranchesList, null, 2));
+    const branchesToCascade = fullBranchesList
+      .filter((branch) => cascadingBranchesPattern.test(branch) || branch === config.defaultBranch)
+      .filter((branch) => !ignoredPatterns.some((pattern) => pattern.test(branch)))
+      .map((branch) => {
+        if (branch === config.defaultBranch) {
+          return {
+            branch,
+            semver: undefined
+          };
+        } else {
+          const version = versionCapturePattern.exec(branch)?.[1];
+          if (!version) {
+            return {
+              branch,
+              semver: null
+            };
+          }
+          const parsedVersion = parse(version);
+          return {
+            branch,
+            semver: parsedVersion ? parsedVersion : parse(valid(coerce(version)))
+          };
+        }
+      })
+      .filter(({branch, semver}) => {
+        if (semver === null) {
+          this.logger.warn(`Failed to parse the branch ${branch}, it will be skipped from cascading`);
+          return false;
+        }
+        return true;
+      })
+      .sort((branchObjectA, branchObjectB) => {
+        if (!branchObjectA.semver) {
+          return 1;
+        } else if (!branchObjectB.semver) {
+          return -1;
+        }
+
+        return compare(branchObjectA.semver, branchObjectA.semver);
```

Comment:
should't be: 
```
        return compare(branchObjectA.semver, branchObjectB.semver);
```
?

Reply:
Waouh, it worked magically :D.
Added UT to more test this part with the fix